### DISCUSSION TITLE
[WASM + WindowsAppSDK] How to persist app state on shutdown?

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#12375] [Q&A]
I have to persist the app state on shutdown by an `async` method.

I ran into [that problem](https://github.com/microsoft/WindowsAppSDK/issues/2398) on Windows, but was able to work around by having the relevant logic in AppWindow.Closing -> no problem there.

On WASM I tried:
- adding it to the `Suspending `event -> does not work, even when using the deferral approach similar to [here](https://platform.uno/docs/articles/features/windows-ui-core-preview.html)
- @MartinZikmund's [approach ](https://blog.mzikmund.com/2018/09/app-close-confirmation-in-uwp/)doesn't work neither on WindowsAppSDK, since `SystemNavigationManagerPreview `is not implemented by UNO WASM

What could I do to persist the app state on browser shutdown or browser tab closing?

**Comment by jeromelaban** [comment]
The "Closing" of an app on the web is very fragile, as the user may close the tab an any time and the lifecycle events may not be call at all. See [this document](https://platform.uno/docs/articles/features/windows-ui-xaml-application.html?q=susp#webassembly) for more details.

In general, it's best to store relevant state at any point in time, so it can be restored when the app is loaded again.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban. In fact I already have the relevant persistance logic in place in the `.Suspending event `which you linked. However, it's not working - I have no idea what I'm doing wrong. The same async method works no problem in other situations.
I also have some timer based logic in place already. 

However, the WASM user experience still is not on par with WindowsAppSDK and forces me to add some explicit "Save workspace" logic on UNO WASM - which is not favorable.

**Reply by jeromelaban** [reply]
The experience in that area cannot be on par with WinAppSDK, as the underlying environment does not allow to do so. The only part that you can do is create a browser javascript popup that will ask the user to cancel and save explicitly, but it has to be synchronous in the `Suspending` event.

**Reply by DierkDroth** [reply]
Hmm ... that might be a viable approach. I will give that a try. Thanks @jeromelaban

**Reply by DierkDroth** [reply]
@jeromelaban AFAIK the .[onbeforeunload](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event) logic is the only way to show a popup on browser 'shutdown' which would privide an option that the user could cancel the process.

However, it does not appear to work. Please see attached repo where I added `onbeforeunload` by 
```
Uno.Foundation.WebAssemblyRuntime.InvokeJS("window.onbeforeunload = function() { return 'Do you want to leave this page?'; };");
```

What am I doing wrong?

[OnBeforeUnload.zip](https://github.com/unoplatform/uno/files/11565221/OnBeforeUnload.zip)
